dist/centrifuge.d.ts

```diff
@@ -0,0 +1,116 @@
+    disconnect(): void;
+    ping(): void;
+    startBatching(): void;
+    stopBatching(): void;
+    startSubscribeBatching(): void;
+    stopSubscribeBatching(): void;
+    subscribe(channel: string, events?: (...args: any[]) => void): Centrifuge.Subscription;
+    subscribe(channel: string, events?: Centrifuge.SubscriptionEvents): Centrifuge.Subscription;
+}
+
+declare namespace Centrifuge {
+    export interface Options {
+        debug?: boolean;
+        sockjs?: any;
+        promise?: any;
+        minRetry?: number;
+        maxRetry?: number;
+        timeout?: number;
+        ping?: boolean;
+        pingInterval?: number;
+        pongWaitTimeout?: number;
+        privateChannelPrefix?: string;
+        onTransportClose?: boolean;
+        sockjsServer?: string | null;
+        sockjsTransports?: string[];
+        refreshEndpoint?: string;
+        refreshHeaders?: object;
+        refreshParams?: object;
+        refreshData?: object;
+        refreshAttempts?: number | null;
+        refreshInterval?: number;
+        onRefreshFailed?: () => void;
+        onRefresh?: (context: object, cb: (resp: any) => void) => void;
+        subscribeEndpoint?: string;
+        subscribeHeaders?: object;
+        subscribeParams?: object;
+        subRefreshInterval?: number;
+        onPrivateSubscribe?: (message: {data: SubscribePrivateMessage}, cb: (resp: any) => void) => void;
+    }
+    export class Subscription extends EventEmitter {
+        constructor(centrifuge: Centrifuge, channel: string, events: (...args: any[]) => void);
+        channel: string;
+        ready(callback: (context: SubscribeSuccessContext) => void, errback: (context: SubscribeErrorContext) => void): void;
+        subscribe(): void;
+        unsubscribe(): void;
+        publish(data: any): Promise<any>;
+        presence(): Promise<any>;
+        presenceStats(): Promise<any>;
+        history(): Promise<any>;
+    }
+    export interface SubscriptionEvents {
+        publish?: (message: PublishMessage) => void;
+        join?: (message: JoinLeaveMessage) => void;
+        leave?: (message: JoinLeaveMessage) => void;
+        subscribe?: (context: SubscribeSuccessContext) => void;
+        error?: (errContext: SubscribeErrorContext) => void;
+        unsubscribe?: (context: UnsubscribeContext) => void;
+    }
+    export interface PublishMessage {
+        uid: string;
+        data: any;
+        client?: string;
+        info?: MessageInfo;
+    }
+    export interface MessageInfo {
+        user? : string;
+        client? : string;
+        default_info?: object;
+        channel_info?: object;
+    }
+    export interface JoinLeaveMessage {
+        info: JoinLeaveMessageInfo;
+    }
+    export interface JoinLeaveMessageInfo {
+        user : string;
+        client : string;
+        conn_info?: object;
+        chan_info?: object;
+    }
+    export interface SubscribePrivateMessage {
+        client: string;
+        channels: string[] | object[];
```

Comment:
Is there any place where we use `object[]` at moment? At first glance I see only `string[]` - i.e. array of string channels. Also I think we should rename this to `SubscribePrivateContext` as it's not a message in general - just to be more semantically correct.

Reply:
[This place is here](https://github.com/centrifugal/centrifuge-js#private-channels-subscription)

Reply:
`SubscribePrivateMessage` renamed to `SubscribePrivateContext`

Reply:
> This place is here

Hmm, I think you not right, that callback can only be called with array of strings:
1) https://github.com/centrifugal/centrifuge-js/blob/2.1.6/src/centrifuge.js#L841
2) https://github.com/centrifugal/centrifuge-js/blob/2.1.6/src/centrifuge.js#L1410

The object in readme you pointed to is JSON response from server.

Reply:
As far as I understand, declarations indicate the type of data that the function accepts or returns, regardless of the data source.
If you leave only string[], the typescript user can never use the data returned by the server.
In any case, you can change the declaration at your discretion.

Reply:
Yes, earlier I did not read the documentation carefully.
Now I have a question. 
```json
{
  "client": "<CLIENT ID>",
  "channels": ["$chan1", "$chan2"]
}
```
according to the documentation this is structure of the request. Therefore, it must be passed as an channel argument of `subscribe(channel: string, events?: (...args: any[]) => void): Centrifuge.Subscription;` ?

For example:
```js
centrifuge.subscribe('{
  "client": "<CLIENT ID>",
  "channels": ["$chan1", "$chan2"]
}', function(message) {
    // process message
});
```
And callback argument `message` will take:
```json
{
  "channels": [
    {
      "channel": "$chan1",
      "token": "<SUBSCRIPTION JWT TOKEN>"
    },
    {
      "channel": "$chan2",
      "token": "<SUBSCRIPTION JWT TOKEN>"
    }
  ]
}
```

Reply:
Do I understand correctly that `SubscribePrivateContext` is used only at `Options.onPrivateSubscribe` method and callback will take server response?
If it is true:
```ts
    export interface SubscribePrivateContext {
        client: string;
        channels: string[];
    }
```
Also need an interface for callback response.

Reply:
When client subscribes on private channel by default AJAX request will be sent to `subscribeEndpoint` automatically when channel starts with `privateChannelPrefix`. It does not involve any code from developer side - everything happens inside centrifuge-js. Developer only needs to return proper response from server in this case.

But there is a way to override default behaviour and take full control on authorizing private channels. In this case developer can provide custom `onPrivateSubscribe` function. It should get `SubscribePrivateContext` as first argument and call callback (it's second argument) when finish authorizing provided channels. It should call callback with data like (i.e. data similar to what must be returned from server in automatic case):

```json
{
  "channels": [
    {
      "channel": "$chan1",
      "token": "<SUBSCRIPTION JWT TOKEN>"
    },
    {
      "channel": "$chan2",
      "token": <SUBSCRIPTION JWT TOKEN>
    }
  ]
}
```

Reply:
Looks like sth like this is correct:

```typescript
export interface SubscribePrivateContext {
        client: string;
        channels: string[];
}

export interface SubscribePrivateResponse {
        channels: PrivateChannelData[];
}

export interface PrivateChannelData {
        channel: string;
        token: string;
}

onPrivateSubscribe?: (message: {data: SubscribePrivateContext}, cb: (resp: SubscribePrivateResponse) => void) => void;
```

Where each object in `SubscribePrivateResponse` is

Reply:
> When client subscribes on private channel by default AJAX request will be sent to `subscribeEndpoint` automatically when channel starts with `privateChannelPrefix`. It does not involve any code from developer side - everything happens inside centrifuge-js. Developer only needs to return proper response from server in this case.
> 
> But there is a way to override default behaviour and take full control on authorizing private channels. In this case developer can provide custom `onPrivateSubscribe` function. It should get `SubscribePrivateContext` as first argument and call callback (it's second argument) when finish authorizing provided channels. It should call callback with data like:

It would be nice to add this to the documentation.

Reply:
> ```ts
> export interface SubscribePrivateContext {
>         client: string;
>         channels: string[];
> }
> 
> export interface SubscribePrivateResponse {
>         channels: PrivateChannelData[];
> }
> 
> export interface PrivateChannelData {
>         channel: string;
>         token: string;
> }
> 
> onPrivateSubscribe?: (message: {data: SubscribePrivateContext}, cb: (resp: SubscribePrivateResponse) => void) => void;
> ```
Yes, it correct

Reply:
Added to docs